torch/utils/_benchmark/utils/valgrind_wrapper.py

```diff
@@ -0,0 +1,227 @@
+import os
+import re
+import shutil
+import subprocess
+import sys
+import tempfile
+import textwrap
+
+import torch
+
+
+class _ValgrindWrapper(object):
+    def __init__(self):
+        self._valgrind_available = not subprocess.run(
+            ["which", "valgrind"], capture_output=True).returncode
+
+        self._build_type = None
+        build_search = re.search("BUILD_TYPE=(.+),", torch.__config__.show())
+        if build_search is not None:
+            self._build_type = build_search.groups()[0].split(",")[0]
+
+        self._baseline_cache = {}
+
+    def _validate(self):
+        if not self._valgrind_available:
+            raise OSError("Could not find `valgrind`")
+
+        if self._build_type is None:
+            raise OSError("Could not determine BUILD_TYPE")
+
+        if self._build_type != "RelWithDebInfo":
+            raise OSError(
+                "PyTorch sould be built with REL_WITH_DEB_INFO=1 "
+                "(debug symbols) in order to produce a useful C++ call graph, "
+                f"got BUILD_TYPE={self._build_type} instead"
+            )
+
+    def collect_callgrind(self, stmt: str, setup: str, number: int, num_threads: int):
+        self._validate()
+        if number not in self._baseline_cache:
+            self._baseline_cache[(number, num_threads)] = self._invoke(
+                stmt="pass", setup="pass", number=number, num_threads=num_threads)
```

Comment:
So _baseline_cache mostly captures some part of `subprocess.run()` spent on communicating with callgrind_control to start/stop the callgrind?

Just curious - have you thought about offering an option to actually run the stmt once in baseline, i.e. comparing (N + 1) times against 1 times. This way it would ignore some static initialization cost (e.g. static PythonArgParser initialization which happens at runtime when you call an op for the first time), which might or might not be relevant, depending on what we are trying to measure.

Reply:
I do several warmup iterations in the `User specified setup` block. The `subprocess.run` blocks were definitely the bulk of the instructions, but even with them gone there are still a couple thousand instructions from various parts of the interpreter. My concern comparing 1 vs. (N + 1) is that you would no longer be able to reuse baselines. But 100% agree that we don't want to benchmark any of the various lazy framework inits.